### ISSUE TITLE
swarm: Make API for creating a new `Swarm` executor aware

### ISSUE THREAD

**Issue by thomaseizinger** [libp2p/rust-libp2p#3068]
## Description

This issue is extracted out of #2173 and proposes a solution to issues like https://github.com/libp2p/rust-libp2p/issues/2230.

A `Swarm` needs to execute background tasks. Currently, each physical connection (i.e. a TCP connection) is executed on its own background task. This improves latency because it means the main event loop of the `Swarm` which also calls `NetworkBehaviour::poll` does not get blocked by IO. Also see https://github.com/libp2p/rust-libp2p/issues/2885.

Today, the way a `Swarm` achieves this is very subtle.

By default, `Swarm::new` will try to create a `futures::executor::Threadpool` to execute these connection background tasks. For async-IO types from the `tokio` runtime, this does not work because they require a tokio reactor to run on the same thread which by default is only active on a worker thread of a tokio `Runtime`. To achieve this, a user has to use the `SwarmBuilder` instead and call `executor(Box::new(|f| { tokio::spawn(f) }))`. See for example here:

https://github.com/libp2p/rust-libp2p/blob/b42f28630e2a7a04642c95aa8e7f4deea823f901/examples/chat-tokio.rs#L109-L115

In case creation of the `futures::executor::Threadpool` fails, the connection tasks are polled on the current thread instead:

https://github.com/libp2p/rust-libp2p/blob/b42f28630e2a7a04642c95aa8e7f4deea823f901/swarm/src/connection/pool.rs#L398-L404

https://github.com/libp2p/rust-libp2p/blob/b42f28630e2a7a04642c95aa8e7f4deea823f901/swarm/src/connection/pool.rs#L101-L103

I am proposed to change the API for `Swarm` and `SwarmBuilder` to force the user to make a decision on which executor they'd like their tasks to be executed on. There are several options and it is not yet clear, what the best one is:

- Introduce a type parameter on `Swarm` that has a trait-bound for spawning new tasks. This would allow us to re-export type-aliases like `libp2p::swarm::tokio::Swarm` which would point to something like `libp2p::swarm::Swarm<TokioRuntime>`. A type-parameter-only solution is likely the easiest to migrate too because it doesn't require any signature changes. We can deprecate the current type in favor of `libp2p::swarm::futures_threadpool::Swarm`.
- Introduce a new parameter to `libp2p::swarm::Swarm::new` that requires to pass an executor-specific type:
   ```rust
   impl Swarm {
       fn new<T>(..., executor: T) -> Self { }
   }

   Swarm::new(..., TokioExecutor);
   ```

There are likely other solutions too. Whatever we choose, it must play nicely with cargo features, i.e. be completely additive in terms of APIs and not change any behaviour.

## Motivation

- Make it easier for users to use `rust-libp2p` correctly.
- Less boilerplate to configure a custom executor.
- Treat all executors equally.

## Are you planning to do it yourself in a pull request?

Maybe.

**Comment by umgefahren** [comment]
I'm currently working on a PR to solve this issue. However, I have a small question: Solving this could remove the need for local spawns since FuturesUnordered could also be treated as an executor. I don't see any usages of local spawns besides being a fallback.

**Comment by thomaseizinger** [comment]
> I'm currently working on a PR to solve this issue. However, I have a small question: Solving this could remove the need for local spawns since FuturesUnordered could also be treated as an executor. I don't see any usages of local spawns besides being a fallback. 

It doesn't quite work unfortunately.
 `FuturesUnordered` requires `&mut self` to spawn new futures and it needs to be polled to make progress. Both of these are not needed typically for the executors we are interested in.

**Comment by umgefahren** [comment]
I see, however, if there is an executor, there is no need for local spawns, right? So I could just construct an enum encapsulating the different cases.

**Comment by thomaseizinger** [comment]
> I see, however, if there is an executor, there is no need for local spawns, right? So I could just construct an enum encapsulating the different cases.

That is correct. It is either one or the other :)

**Comment by umgefahren** [comment]
This issue is closed by #3097. Consider linking it.

**Comment by thomaseizinger** [comment]
> This issue is closed by #3097. Consider linking it.

Done!
